### PR TITLE
Flow Thread.CurrentPrincipal with ExecutionContext

### PR DESCRIPTION
**Pull request by MarcoRossignoli**
fixes https://github.com/dotnet/corefx/issues/34489

/cc @jkotas @stephentoub

### CODE REVIEW

**Comment by benaadams** [comment]
Never actually reads from `s_asyncLocalPrincipal`?

**Comment by MarcoRossignoli** [comment]
I used similar code of CurrentCulture https://github.com/dotnet/corefx/blob/5e3fe25d1236cd0192dd966f7286fe008d5cf875/src/Common/src/CoreLib/System/Globalization/CultureInfo.cs#L125

**Comment by benaadams** [comment]
Ah, missed it was being set with the callback

**Comment by stephentoub** [comment]
> Ohhh first CI with Azure Pipeline for me

Before you get too excited, note that it's broken :)

**Comment by MarcoRossignoli** [comment]
test this please

File: src/System.Threading.Thread/src/System/Threading/Thread.cs [hunk 1]
```diff
@@ -190,14 +191,44 @@ public static IPrincipal CurrentPrincipal
...
+                return CurrentThread._principal;
             }
             set
             {
-                CurrentThread._principal = value;
+                SetAsyncLocalPrincipal(value);
             }
         }
 
+        private static void SetAsyncLocalPrincipal(IPrincipal principal)
```
Comment (jkotas) [inline src/System.Threading.Thread/src/System/Threading/Thread.cs]: This should not need to be separate method. The getter implementation above can do `Principal = principal;` to call this.

File: src/System.Threading.Thread/src/System/Threading/Thread.cs [hunk 2]
```diff
@@ -21,6 +21,7 @@ public sealed partial class Thread : CriticalFinalizerObject
         private CultureInfo _startCulture;
         private CultureInfo _startUICulture;
         private IPrincipal _principal;
```
Comment (jkotas) [inline src/System.Threading.Thread/src/System/Threading/Thread.cs]: (Not related to your change) This may look better as ` [ThreadStatic]` - saves fetching the current thread on access.

File: src/System.Threading.Thread/src/System/Threading/Thread.cs [hunk 3]
```diff
@@ -190,14 +191,44 @@ public static IPrincipal CurrentPrincipal
...
             {
-                return CurrentThread._principal ?? (CurrentThread._principal = AppDomain.CurrentDomain.GetThreadPrincipal());
+                if (CurrentThread._principal is null)
+                {
+                    IPrincipal principal = AppDomain.CurrentDomain.GetThreadPrincipal();
+                    if (principal is null)
+                    {
+                        return null;
+                    }
+                    else
```
Comment (jkotas) [inline src/System.Threading.Thread/src/System/Threading/Thread.cs]: Nit: May look better without `else` so that it looks same as the setter.

File: src/System.Threading.Thread/src/System/Threading/Thread.cs [hunk 4]
```diff
@@ -190,14 +191,44 @@ public static IPrincipal CurrentPrincipal
         {
             get
             {
-                return CurrentThread._principal ?? (CurrentThread._principal = AppDomain.CurrentDomain.GetThreadPrincipal());
+                if (CurrentThread._principal is null)
+                {
+                    IPrincipal principal = AppDomain.CurrentDomain.GetThreadPrincipal();
+                    if (principal is null)
```
Comment (jkotas) [inline src/System.Threading.Thread/src/System/Threading/Thread.cs]: I am wondering whether these null shortcuts can cause problems. I cannot think about anything, just pointing it out.

File: src/System.Threading.Thread/tests/ThreadTests.cs [hunk 5]
```diff
@@ -427,6 +429,54 @@ public static void CurrentPrincipalTest()
...
+        [Fact]
+        public static void CurrentPrincipalContextFlowTest_NotFlow()
+        {
+            ThreadTestHelpers.RunTestInBackgroundThread(async () =>
+            {
+                Thread.CurrentPrincipal = new ClaimsPrincipal();
+
+                ExecutionContext.SuppressFlow();
+
+                await Task.Run(() => Assert.Null(Thread.CurrentPrincipal));
```
Comment (jkotas) [inline src/System.Threading.Thread/tests/ThreadTests.cs]: This may be a non-reliable tests. Suppressing the execution flow will mean that the thread will get the execution context from whatever was running on the thread before.

**Comment by jkotas** [comment]
LGTM otherwise

File: src/System.Threading.Thread/src/System/Threading/Thread.cs (hunk 4 above)
Comment (MarcoRossignoli) [reply src/System.Threading.Thread/src/System/Threading/Thread.cs]: On issue you specified to be "lazy" on AsyncLocal instancing https://github.com/dotnet/corefx/issues/34489#issuecomment-453171829 and allocate only on set, so my idea is to avoid allocation for "default" .net core policy get, I don't see particular issue to return null, I think will be used to compare with something else.
BTW you "feeling" take precedence over my simple usage expectations so if you're more confortable with simple assignment to `CurrentPrincipal` I won't oppose resistance :smiley:

Comment (MarcoRossignoli) [reply src/System.Threading.Thread/src/System/Threading/Thread.cs]: Tried different way, similar to previous, should be lazy as possible.

File: src/System.Threading.Thread/src/System/Threading/Thread.cs [hunk 6]
```diff
@@ -16,11 +16,14 @@ public sealed partial class Thread : CriticalFinalizerObject
         [ThreadStatic]
         private static Thread t_currentThread;
 
+        [ThreadStatic]
+        private static IPrincipal s_principal;
```
Comment (jkotas) [inline src/System.Threading.Thread/src/System/Threading/Thread.cs]: This should have `t_` prefix. Also, I would put the AsyncLocal and this static next to each other since they logically belong together.

File: src/System.Threading.Thread/src/System/Threading/Thread.cs (hunk 4 above)
Comment (jkotas) [reply src/System.Threading.Thread/src/System/Threading/Thread.cs]: I was concerned about behavior differences between .NET Core and .NET Framework. The .NET Framework implementation does not have a null check like this and that is what made me think about this. The restate my question: Are there cases where this null check is going cause observable behavior differences between .NET Framework and .NET Core? (As I have said, I cannot think about any.)

Wrt. lazy: My primary concern was about creating the AsyncLocal only once somebody calls the CurrentPrincipal property.

**Comment by jkotas** [comment]
`CurrentPrincipalContextFlowTest_NotFlow` test is failing on Linux.

File: src/System.Threading.Thread/src/System/Threading/Thread.cs (hunk 6 above)
Comment (jkotas) [reply src/System.Threading.Thread/src/System/Threading/Thread.cs]: Right: https://github.com/dotnet/corefx/blob/master/Documentation/coding-guidelines/coding-style.md

**Comment by MarcoRossignoli** [comment]
> CurrentPrincipalContextFlowTest_NotFlow test is failing on Linux.

I'll verify...also on Windows.10.Amd64.ClientRS4.Open-x86-Release

File: src/System.Threading.Thread/src/System/Threading/Thread.cs (hunk 4 above)
Comment (MarcoRossignoli) [reply src/System.Threading.Thread/src/System/Threading/Thread.cs]: Now the only shortcut is for set `null` without previous set something, do you feel ok with it?

File: src/System.Threading.Thread/src/System/Threading/Thread.cs (hunk 6 above)
Comment (MarcoRossignoli) [reply src/System.Threading.Thread/src/System/Threading/Thread.cs]: first use of eyes...it's not clear to me when use it maybe "I'll take a look"  :smile:

File: src/System.Threading.Thread/src/System/Threading/Thread.cs (hunk 4 above)
Comment (jkotas) [reply src/System.Threading.Thread/src/System/Threading/Thread.cs]: Yes, I think it is ok based on the codereview. I assume that you have done enough testing to convince yourself that it behaves the same way as .NET Framework.

File: src/System.Threading.Thread/tests/ThreadTests.cs [hunk 7]
```diff
@@ -427,6 +429,58 @@ public static void CurrentPrincipalTest()
...
+                // TaskCreationOptions.LongRunning to have fresh new thread with clean execution context
+                await Task.Factory.StartNew(() => Assert.IsType<ClaimsPrincipal>(Thread.CurrentPrincipal), CancellationToken.None, TaskCreationOptions.LongRunning, TaskScheduler.Default);
+
+                ExecutionContext.SuppressFlow();
+
+                // TaskCreationOptions.LongRunning to have fresh new thread with clean execution context
+                await Task.Factory.StartNew(() => Assert.Null(Thread.CurrentPrincipal), CancellationToken.None, TaskCreationOptions.LongRunning, TaskScheduler.Default);
+
+                // defaults are null for netcorapp and GenericPrincipal for netfx
+                Assert.True(Thread.CurrentPrincipal is null || Thread.CurrentPrincipal is GenericPrincipal);
```
Comment (stephentoub) [inline src/System.Threading.Thread/tests/ThreadTests.cs]: Hmm, I don't know if this is going to behave the way you're hoping; I'm a little worried this may result in sporadic test failures (I could be wrong). With flow suppressed, I suspect the async local's changed callback won't be invoked, which means this thread's t_principal won't be set to null... if this continuation then ends up running on the same thread that was running previously running it, it'll end up seeing CurrentPrincipal as a ClaimsPrincipal rather than null or GenericPrincipal.

Comment (MarcoRossignoli) [reply src/System.Threading.Thread/tests/ThreadTests.cs]: I believe that your suspect is correct...this could be the reason of ubuntu fail, I would try with custom task factory(new thread for every await) to be sure to have clean thread. About it I've got a question, you state `if this continuation then ends up running on the same thread that was running previously running it` this mean that there isn't context/tls cleanup when a thread returns to pool?If so, isn't this dangerous?

Comment (stephentoub) [reply src/System.Threading.Thread/tests/ThreadTests.cs]: > this mean that there isn't context/tls cleanup when a thread returns to pool?If so, isn't this dangerous?

Thread locals are not reset for every work item on a thread pool thread.  That would lead to really bad performance in common cases, e.g. where you cache things in a ThreadStatic, as is done in ArrayPool, or StringBuilderCache, or other such places.

Comment (MarcoRossignoli) [reply src/System.Threading.Thread/tests/ThreadTests.cs]: understood thank's.

Comment (MarcoRossignoli) [reply src/System.Threading.Thread/tests/ThreadTests.cs]:
```cs
private static async Task Test()
{
    Thread.CurrentPrincipal = new ClaimsPrincipal();

    ExecutionContext.SuppressFlow();

    await Task.Factory.StartNew(() => Console.WriteLine(Thread.CurrentPrincipal), CancellationToken.None, TaskCreationOptions.LongRunning, TaskScheduler.Default);

    await Task.Run(() => Console.WriteLine(Thread.CurrentPrincipal));
}
...
.NET Full C:\Windows\Microsoft.NET\Framework\v4.0.30319\mscorlib.dll
System.Security.Claims.ClaimsPrincipal
System.Security.Principal.GenericPrincipal
```
On full framework suppress flow doesn't work for `StartNew` is it expected?

**Comment by stephentoub** [comment]
> On full framework suppress flow doesn't work for StartNew is it expected?

This isn't about StartNew: it's about Thread.Start.  Thread.Start on netfx has odd behavior with regards to principal, flowing it separate from ExecutionContext.
https://referencesource.microsoft.com/#mscorlib/system/threading/thread.cs,353
You can see the relevant native side of this in coreclr history:
https://github.com/dotnet/coreclr/blob/ef1e2ab328087c61a6878c1e84f4fc5d710aebce/src/vm/comsynchronizable.cpp#L439

File: src/System.Threading.Thread/tests/ThreadTests.cs (hunk 7 above)
Comment (MarcoRossignoli) [reply src/System.Threading.Thread/tests/ThreadTests.cs]: Ah good to know thanks's for the infos, I removed test from netfx because to avoid issue with tls I need a new thread, I removed also second check, it doesn't have a lot of sense we neeed only to check if context is not flowed not what is context after.

**Comment by MarcoRossignoli** [comment]
test Linux arm64 Release Build please
test corefx-ci please

File: src/System.Threading.Thread/tests/ThreadTests.cs [hunk 8]
```diff
@@ -427,6 +429,56 @@ public static void CurrentPrincipalTest()
...
+                Thread.CurrentPrincipal = new ClaimsPrincipal();
+
+                // TaskCreationOptions.LongRunning to have fresh new thread with clean execution context
+                await Task.Factory.StartNew(() => Assert.IsType<ClaimsPrincipal>(Thread.CurrentPrincipal), CancellationToken.None, TaskCreationOptions.LongRunning, TaskScheduler.Default);
+
+                ExecutionContext.SuppressFlow();
+
+                // TaskCreationOptions.LongRunning to have fresh new thread with clean execution context
+                await Task.Factory.StartNew(() => Assert.True(Thread.CurrentPrincipal is null), CancellationToken.None, TaskCreationOptions.LongRunning, TaskScheduler.Default);
+            });
```
Comment (danmosemsft) [inline src/System.Threading.Thread/tests/ThreadTests.cs]: Should we test that re-enabling flow works? From the docs seems like that's `ExecutionContext.SuppressFlow().Undo()` ?

Comment (MarcoRossignoli) [reply src/System.Threading.Thread/tests/ThreadTests.cs]: You're right, however we cannot undo because we get exception, undo must be called on thread called suppress(continuation is on other thread pool thread), so I preferred avoid risks and run test in custom process, is it a good idea?

Comment (stephentoub) [reply src/System.Threading.Thread/tests/ThreadTests.cs]: > however we cannot undo because we get exception

You can, you just can't around an await, so instead of:
```C#
using (ExecutionContext.SuppressFlow())
{
    await Task.*;
}
```
you'd do:
```C#
Task t;
using (ExecutionContext.SuppressFlow())
{
    t = Task.*;
}
await t;
```

Comment (MarcoRossignoli) [reply src/System.Threading.Thread/tests/ThreadTests.cs]: Good point!

**Comment by MarcoRossignoli** [comment]
test Linux arm64 Release Build please

**Comment by MarcoRossignoli** [comment]
Issue with Linux CI, expect at moment? @danmosemsft @stephentoub @jkotas

**Comment by MarcoRossignoli** [comment]
Guys, unrelated question, can you explain to me what is Helix?I see this name everywhere but nothing on guides, is this MS internal testing tool?

**Comment by danmosemsft** [comment]
Its useful because it is massively parallel...it sends every test library to a different  azure machine so all the tests complete very quickly

**Comment by MarcoRossignoli** [comment]
Thank's a lot for the info!I regarding this PR O don't know if needs to have green CI before merge, do we wait Linux CI fix?

**Comment by MarcoRossignoli** [comment]
@jkotas PTAL

**Comment by MarcoRossignoli** [comment]
@safern Will Jenkins removed from CI?If so is there a date?

**Comment by safern** [comment]
> @safern Will Jenkins removed from CI?If so is there a date?

Yes it will be. There is not a definite date, but should be removed soon as it seems like the azure devops leg is stabilized.

File: src/System.Threading.Thread/src/System/Threading/Thread.cs [hunk 9]
```diff
@@ -190,14 +193,31 @@ public static IPrincipal CurrentPrincipal
         {
             get
             {
-                return CurrentThread._principal ?? (CurrentThread._principal = AppDomain.CurrentDomain.GetThreadPrincipal());
+                if (t_principal is null)
+                {
+                   CurrentPrincipal = AppDomain.CurrentDomain.GetThreadPrincipal();
```
Comment (stephentoub) [inline src/System.Threading.Thread/src/System/Threading/Thread.cs]: Nit: looks like the indentation here is slightly off

File: src/System.Threading.Thread/src/System/Threading/Thread.cs [hunk 10]
```diff
@@ -190,14 +193,31 @@ public static IPrincipal CurrentPrincipal
...
-                CurrentThread._principal = value;
+                if (s_asyncLocalPrincipal is null)
+                {
+                    if (value is null)
+                    {
+                        return;
+                    }
+                    Interlocked.CompareExchange(ref s_asyncLocalPrincipal, new AsyncLocal<IPrincipal>(CurrentPrincipalContextChanged), null);
+                }
+                s_asyncLocalPrincipal.Value = value;
```
Comment (stephentoub) [inline src/System.Threading.Thread/src/System/Threading/Thread.cs]: The code here is indirectly setting t_principal, but it's not obvious from the code.  Can we add an assert, something like:
```C#
Debug.Assert(value == null || t_principal != null);
```
?

File: src/System.Threading.Thread/src/System/Threading/Thread.cs [hunk 11]
```diff
@@ -16,11 +16,14 @@ public sealed partial class Thread : CriticalFinalizerObject
         [ThreadStatic]
         private static Thread t_currentThread;
 
+        [ThreadStatic]
+        private static IPrincipal t_principal;
+        private static AsyncLocal<IPrincipal> s_asyncLocalPrincipal;
```
Comment (stephentoub) [inline src/System.Threading.Thread/src/System/Threading/Thread.cs]: @jkotas, I think you'd requested this split?  Was that for performance reasons or for functional behavior?  If the latter, ok.  If the former, I wonder if we'd be better served just having s_asyncLocalPrincipal, which could still be lazily initialized.  It's a little hard to follow the logic about how/when t_principal gets set, and that would be cleaned up if we didn't have it at all.

File: src/System.Threading.Thread/tests/ThreadTests.cs [hunk 12]
```diff
@@ -427,6 +429,89 @@ public static void CurrentPrincipalTest()
...
+
+                Task task;
+                using(ExecutionContext.SuppressFlow())
+                {
+                    Assert.True(ExecutionContext.IsFlowSuppressed());
+
+                    task = Task.Run(() => 
+                    {
+                        // Default PrincipalPolicy for netcoreapp is null and for netfx is ClaimsPrincipal
+                        Assert.True(Thread.CurrentPrincipal is null || Thread.CurrentPrincipal is ClaimsPrincipal);
```
Comment (stephentoub) [inline src/System.Threading.Thread/tests/ThreadTests.cs]: This is going to pass on netcoreapp even if we have the wrong behavior of the principal flowing.

I suggest just making the test SkipOnTargetFramework, skipping it for netfx, and then making this logic as tight as we can to correctly validate the behavior we expect on netcoreapp.

File: src/System.Threading.Thread/tests/ThreadTests.cs [hunk 13]
```diff
@@ -427,6 +429,89 @@ public static void CurrentPrincipalTest()
...
+                await task;
+            });
+        }
+
+        [Fact]
+        public static void CurrentPrincipal_SetNull()
+        {
+            // We run test on remote process because we need to set same principal policy
+            // On netfx default principal policy is PrincipalPolicy.UnauthenticatedPrincipal
+            DummyClass.RemoteInvoke(() =>
```
Comment (stephentoub) [inline src/System.Threading.Thread/tests/ThreadTests.cs]: I only just noticed this `DummyClass` thing.  I realize you're just following along with the rest of the tests, so nothing for you to changed, but I'm curious why this separate type is needed.

File: src/System.Threading.Thread/src/System/Threading/Thread.cs (hunk 11 above)
Comment (MarcoRossignoli) [reply src/System.Threading.Thread/src/System/Threading/Thread.cs]: It wasn't requested by @jkotas I saw this on CurrentCulture and I thought to use same "style" https://github.com/dotnet/corefx/blob/5e3fe25d1236cd0192dd966f7286fe008d5cf875/src/Common/src/CoreLib/System/Globalization/CultureInfo.cs#L125

File: src/System.Threading.Thread/tests/ThreadTests.cs (hunk 12 above)
Comment (MarcoRossignoli) [reply src/System.Threading.Thread/tests/ThreadTests.cs]: It was a request https://github.com/dotnet/corefx/pull/34747#discussion_r250276887, to be sure that the behaviour is the same. Maybe I could fix asserts with https://github.com/dotnet/corefx/blob/master/src/CoreFx.Private.TestUtilities/src/System/PlatformDetection.cs#L24

File: src/System.Threading.Thread/tests/ThreadTests.cs (hunk 13 above)
Comment (MarcoRossignoli) [reply src/System.Threading.Thread/tests/ThreadTests.cs]: I saw, I followed other tests style, do i fix?

File: src/System.Threading.Thread/src/System/Threading/Thread.cs (hunk 10 above)
Comment (MarcoRossignoli) [reply src/System.Threading.Thread/src/System/Threading/Thread.cs]: ok

File: src/System.Threading.Thread/tests/ThreadTests.cs (hunk 12 above)
Comment (stephentoub) [reply src/System.Threading.Thread/tests/ThreadTests.cs]: > It was a request #34747 (comment), to be sure that the behaviour is the same

Yes, but it's no longer validating that the behavior is the same.

Comment (MarcoRossignoli) [reply src/System.Threading.Thread/tests/ThreadTests.cs]: yes I agree test is wrong right now, I'll fix with `PlatformDetection` helper.

File: src/System.Threading.Thread/tests/ThreadTests.cs (hunk 13 above)
Comment (stephentoub) [reply src/System.Threading.Thread/tests/ThreadTests.cs]: No, you can leave it as is.

File: src/System.Threading.Thread/src/System/Threading/Thread.cs (hunk 11 above)
Comment (jkotas) [reply src/System.Threading.Thread/src/System/Threading/Thread.cs]: The Principal property should not be used on performance critical paths. It should be ok to not cache it in a thread local, and always go through async-local getter.

Comment (MarcoRossignoli) [reply src/System.Threading.Thread/src/System/Threading/Thread.cs]: ok I'll update implementation with direct access to `AsyncLocal` value getter, thank's!

File: src/System.Threading.Thread/tests/ThreadTests.cs (hunk 12 above)
Comment (MarcoRossignoli) [reply src/System.Threading.Thread/tests/ThreadTests.cs]: fixed

File: src/System.Threading.Thread/src/System/Threading/Thread.cs (hunk 11 above)
Comment (MarcoRossignoli) [reply src/System.Threading.Thread/src/System/Threading/Thread.cs]: done

File: src/System.Threading.Thread/src/System/Threading/Thread.cs (hunk 9 above)
Comment (MarcoRossignoli) [reply src/System.Threading.Thread/src/System/Threading/Thread.cs]: fixed

**Comment by danmosemsft** [comment]
@safern Azure CI failed due to "infrastructure issue". Can you remind me how to invoke it again?

**Comment by safern** [comment]
Yeah. This was because of the DNS outtage. 

/AzurePipelines run corefx-ci

However comment triggers are broken due to a github bug apparently. I spoke with azure people and they are aware, will let me know whenever fixed. I’ve retried the attempt in the build UI on the right top corner there are 3 dots, that open a submenu with the retry option.

**Comment by MarcoRossignoli** [comment]
> I’ve retried the attempt in the build UI on the right top corner there are 3 dots, that open a submenu with the retry option.

@safern  I don't see that menù is only for "MS internals"? (I see "Download the log" or "View queue time.../Add tags"

**Comment by stephentoub** [comment]
Thanks!

**Comment by safern** [comment]
> @safern I don't see that menù is only for "MS internals"? (I see "Download the log" or "View queue time.../Add tags"

Unfortunately now only MS people can use the retry feature. We've given feedback to be able to retry through comments as that would allow externals to do so. I'll communicate it whenever it is ready 😄

**Comment by danmosemsft** [comment]
Thanks for doing this @MarcoRossignoli , it was a journey 😄
